lib/ClangImporter/ClangImporter.cpp

```diff
@@ -7471,6 +7471,75 @@ bool importer::isForeignReferenceTypeWithoutImmortalAttrs(const clang::QualType
          !hasImmortalAtts(pointeeType->getDecl());
 }
 
+llvm::SmallVector<const clang::RecordDecl *, 4>
+getMatchingParentDecls(const clang::RecordDecl *decl) {
+  assert(decl && "decl is null inside getMatchingParentDecls");
+
+  llvm::SmallVector<const clang::RecordDecl *, 4> matchingDecls;
+
+  // Base case
+  if (hasImportAsRefAttr(decl)) {
+    matchingDecls.push_back(decl);
+  }
+
+  if (const auto *cxxRecordDecl = llvm::dyn_cast<clang::CXXRecordDecl>(decl)) {
+    if (!cxxRecordDecl->hasDefinition())
+      return matchingDecls;
+    cxxRecordDecl->forallBases([&](const clang::CXXRecordDecl *baseDecl) {
+      if (baseDecl->hasDefinition() && hasImportAsRefAttr(baseDecl)) {
+        matchingDecls.push_back(baseDecl);
+      }
+      return true;
+    });
+  }
+
+  return matchingDecls;
+}
+
+const clang::RecordDecl *getRefParentOrDiag(const clang::RecordDecl *decl,
+                                            ASTContext &ctx) {
+  auto refParentDecls = getMatchingParentDecls(decl);
+  if (!refParentDecls.empty()) {
+    llvm::StringRef uniqueRetainAttr, uniqueReleaseAttr;
+    for (const auto *refParentDecl : refParentDecls) {
+      assert(refParentDecl &&
+             "refParentDecl is null inside getRefParentOrDiag");
+      for (const auto *attr : refParentDecl->getAttrs()) {
+        if (const auto swiftAttr = llvm::dyn_cast<clang::SwiftAttrAttr>(attr)) {
+          const auto &attribute = swiftAttr->getAttribute();
+          if (attribute.starts_with("retain:")) {
+            if (uniqueRetainAttr.empty()) {
```

Comment:
What would happen if we encountered attributes in this order:
```
retain:foo
retain:bar
retain:foo
 ```
 ?
 
 I suspect that it is possible that in hte second iteration we set the string to empty, and in the 3rd one fill it again. I think this is not what we want here. After the first mismatch we should emit a diagnostic and give up further processing.

Reply:
Good catch! I should break after the first mismatch.